llama/runner/runner.go

```diff
@@ -112,21 +109,15 @@ func (s *Server) NewSequence(prompt string, images []ImageData, params NewSequen
 		params.numKeep = len(inputs)
 	}
 
-	if !params.embedding {
-		// Subtracting 4 ensures that at least 1 input can be discarded during shift
-		params.numKeep = min(params.numKeep, s.cache.numCtx-4)
-		params.numKeep += s.bosToken
-	} else {
-		// Embeddings are 1 shot - just truncate to the context window, without ever shifting
-		params.numKeep = min(params.numKeep, s.cache.numCtx)
+	if s.model.AddBOSToken() {
```

Comment:
Thanks for handling this!

This isn't a review comment, but more of an observation: at some point (for future models at least) we should be able to require the BOS token be in the template to avoid needing this "add bos/eos token" logic which causes issues in several places (especially given model metadata doesn't always have this value set correctly).

Reply:
I agree that it would be better to have a single place to handle BOS tokens going forward, the current way is confusing and error prone.

We are handling it currently (at least to the extent in this patch), the change here is just trying to things more consistent and handled in a single place in the runner to reduce bugs.